### DISCUSSION TITLE
Support for bindparam() in session inserts.

### DISCUSSION THREAD

**Discussion by jonjonw** [sqlalchemy/sqlalchemy#9581] [Usage Questions] [answered]
I'm not sure if this is a bug or a request for a new feature but it would be great if I could use bindparam() with ORM session inserts.

For example, bindparam() with session and select works
```
stmt = (
    select(model.Atc).
    filter_by(code=bindparam('atc_code')))
result = session.scalar(stmt, dict(atc_code='bar'))
print(result)  # Atc(code=bar, name=bar-name, id=1)
```

and bindparam() with core insert works
```
stmt = (
    insert(model.Atc).
    values(dict(name=bindparam("bound"))))
session.connection().execute(
    stmt,
    dict(code="test", bound="Frombound"))
result = session.scalar(select(model.TransactionType).filter_by(code="test"))
print(result)  # Atc(code=test, name=Frombound, id=2)
```

but bindparam() with session and insert fails
```
session.execute(
    stmt,
    dict(code="test2", bound="Frombound"))

Exception has occurred: StatementError
(sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'bound'
[SQL: INSERT INTO atc (code, name) VALUES (%(code)s::VARCHAR, %(bound)s::VARCHAR) RETURNING atc.id]
[parameters: [{'code': 'test2'}]]
(Background on this error at: https://sqlalche.me/e/20/cd3x)
sqlalchemy.exc.InvalidRequestError: A value is required for bind parameter 'bound' (Background on this error at: https://sqlalche.me/e/20/cd3x)
```

It seems to swallow the bound parameter but not bind it. I'm not sure if this is a bug or just not a feature of the session.

**Comment by CaselIT** [comment]
Hi,

This seems like a bug.
Providing only bound in the execute dict works? Or alternatively adding both code and bound as bind parameters in the insert changes things?

**Comment by zzzeek** [comment]
that would be a bug and is likely a regression since 2.0 added the ORM insert thing, can I get a hint why you need bindparam() to be placed in this way?   You can pass the parameter in any case without specifying it in values and it will be part of the INSERT, or you can name the bindparam the same as the column to follow the existing convention for INSERT.

**Reply by zzzeek** [reply]
added in #9583

**Reply by jonjonw** [reply]
Sure, it was popping up for me in a few ways and I just tried to boil it down to the simplest case. One example was attempting to use scalar_subquery():

```
    atc_ssq = (
        select(model.AccountTaxClass.id).
        filter_by(code=bindparam("tax_code")).
        scalar_subquery()
    )
    upsert_account = pg_upsert(model.Account).values(tax_id=atc_ssq)
    upsert_account = upsert_account.on_conflict_do_update(
        constraint="account_code_key",
        set_={k: v
              for k, v in upsert_account.excluded.items()
              if k not in ('id', 'code', 'created_at', 'updated_at')}
    )

        accounts = session.scalars(upsert_account.returning(model.Account),
                                   accounts_data,
                                   execution_options={"populate_existing": True})
```

Where AccountTaxClass.code has a unique constraint and accounts_data is read from a yaml file:

```
- code: test1
  tax_code: A
  name: nameA

- code: test2
  tax_code: B
  name: nameB

```

**Reply by zzzeek** [reply]
oh, OK right off, that's got other things going on that are not in the bug I fixed.   The above can work with "raw" and in 2.0.9 with "orm" mode, but I'm not sure it would work with the default of "auto" because that's going to drop into "bulk" again and im not sure if that logic can handle the on conflict stuff.  need to add a test to see what happens.

**Reply by zzzeek** [reply]
it works, we're good

**Reply by zzzeek** [reply]
there was a test for this case in fact, just not with the bindparam() part, added that in.

**Comment by zzzeek** [comment, marked as answer]
OK this can't run the way it's given, because the ORM is attempting to autodetect if this is a "bulk" insert or not based on params.   you can run the statement like this:

```py
session.execute(
    stmt.execution_options(dml_strategy="raw"),
    dict(code="test2", bound="Frombound"))
)
```

since this is an unusual case I'm considering two options.  one is we raise an error when this is detected and ask that "raw" is passed.  the other is we try to support params with ORM mode inserts but I need to see what's involved there.

**Reply by zzzeek** [reply]
OK I think I will be able to support the case fully.   the above workaround will work for now.
